### PR TITLE
scikit-learn-compatible API

### LINKED ISSUE #53: Use OpenML and `sklearn.datasets.fetch_openml` for datasets.
Right now the datasets are expected to be manually downloaded by the user and put in the right place.

They can be uploaded to openml (https://www.openml.org/) and retrieved by [`sklearn.datasets.fetch_openml()`](https://scikit-learn.org/dev/modules/generated/sklearn.datasets.fetch_openml.html#sklearn.datasets.fetch_openml) to avoid having to have the user manually download the datasets.

### PR DESCRIPTION
**Pull request by hoffmansc**
Progress towards #58 

Old API:
* Minor changes to old API mostly to aid reproducibility.
* Fixed bug in CalibratedEqOddsPostprocessing -- GFNR is weighted by base_rate not 1-base_rate
* <details><summary><b>New Sphinx docs layout</b></summary><p>
  new docs pages
  </p></details>

New API:
* **4 datasets** (Adult, German, Bank, Compas) in DataFrame format with protected attributes in the index
  * Automatically downloads from openml.org -- closes #53
* **6 group fairness metrics** as functions (`statistical_parity_difference`, `disparate_impact_ratio`, `equal_opportunity_difference`, `average_odds_difference`, `average_odds_error`, `between_group_generalized_entropy_error`)
* **2 individual fairness metrics** as functions (`generalized_entropy_index` and its variants, `consistency_score`)
* **5 additional metrics** as functions (`specificity_score`, `base_rate`, `selection_rate`, `generalized_fpr`, `generalized_fnr`)
* **3 algorithms** (`Reweighing`, `AdversarialDebiasing`, `CalibratedEqualizedOdds`)

### CODE REVIEW

**Comment by animeshsingh** [comment]
Thanks @hoffmansc. Maybe you can schedule 30 min review with @Tomcli myself ?

**Comment by hoffmansc** [comment]
Sure. Sent an invite.

File: aif360/algorithms/postprocessing/calibrated_eq_odds_postprocessing.py [hunk 1]
```diff
@@ -208,4 +208,4 @@ def weighted_cost(fp_rate, fn_rate, cm, privileged):
             * (1 - cm.base_rate(privileged=privileged))) +
            (fn_rate / norm_const
             * cm.generalized_false_negative_rate(privileged=privileged)
-            * (1 - cm.base_rate(privileged=privileged))))
+            * cm.base_rate(privileged=privileged)))
```
Comment (nrkarthikeyan) [inline aif360/algorithms/postprocessing/calibrated_eq_odds_postprocessing.py]: looks good

File: aif360/algorithms/inprocessing/adversarial_debiasing.py [hunk 2]
```diff
@@ -177,14 +179,15 @@ def fit(self, dataset):
...
+                with tf.control_dependencies([classifier_minimizer]):
+                    adversary_minimizer = adversary_opt.minimize(pred_protected_attributes_loss, var_list=adversary_vars)#, global_step=global_step)
 
             self.sess.run(tf.global_variables_initializer())
             self.sess.run(tf.local_variables_initializer())
 
             # Begin training
             for epoch in range(self.num_epochs):
-                shuffled_ids = np.random.choice(num_train_samples, num_train_samples)
+                shuffled_ids = np.random.choice(num_train_samples, num_train_samples, replace=False)
```
Comment (nrkarthikeyan) [inline aif360/algorithms/inprocessing/adversarial_debiasing.py]: I guess the changes to this function are to aid reproducibility? The replace=False seems to ensure that the batches will be mutually exclusive.

Comment (hoffmansc) [reply aif360/algorithms/inprocessing/adversarial_debiasing.py]: Yes, the changes above were to aid reproducibility. `replace=False` will choose without replacement so the batches will be mutually exclusive, yes. That's more common for mini-batches so I figured it was an oversight although it works fine either way. @pronics2004, can you chime in if there was a reason you did it with replacement originally?

File: tests/sklearn/test_adversarial_debiasing.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
+import numpy as np
```
Comment (nrkarthikeyan) [inline tests/sklearn/test_adversarial_debiasing.py]: Could you add some simple comments to the test_*.py files? Just a couple of lines that discuss some non-obvious (if at all there is any) things should be fine.

File: aif360/datasets/adult_dataset.py [hunk 4]
```diff
@@ -99,7 +99,7 @@ def __init__(self, label_name='income-per-year',
             import sys
             sys.exit(1)
 
-        df = pd.concat([train, test], ignore_index=True)
+        df = pd.concat([test, train], ignore_index=True)
```
[the remark this answers is not in the document]
Comment (hoffmansc) [reply aif360/datasets/adult_dataset.py]: For some reason it's reversed on openml. I could've equivalently changed the new function to cut and paste the indexes correctly but there was no technical reason to do so, just to make it match the previous function and this seemed easier.

File: aif360/sklearn/datasets/openml_datasets.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        >>> adult_num = fetch_adult(numeric_only=True)
+        >>> adult_num.X.shape
+        (48842, 5)
+    """
+    if subset not in {'train', 'test', 'all'}:
+        raise ValueError("subset must be either 'train', 'test', or 'all'; "
+                         "cannot be {}".format(subset))
+    df = to_dataframe(fetch_openml(data_id=1590, target_column=None,
+                                   data_home=data_home or DATA_HOME_DEFAULT))
+    if subset == 'train':
```
[the remark this answers is not in the document]
Comment (hoffmansc) [reply aif360/sklearn/datasets/openml_datasets.py]: Yes, the dataset is distributed as two separate files and most papers use the explicit test set.

File: aif360/sklearn/metrics/metrics.py [hunk 6]
```diff
@@ -0,0 +1,585 @@
...
+def statistical_parity_difference(*y, prot_attr=None, priv_group=1, pos_label=1,
+                                  sample_weight=None):
+    r"""Difference in selection rates.
+
+    .. math::
+        Pr(\hat{Y} = \text{pos_label} | D = \text{unprivileged})
+        - Pr(\hat{Y} = \text{pos_label} | D = \text{privileged})
+
+    Note:
+        If only y_true is provided, this will return the difference in base
```
[the remark this answers is not in the document]
Comment (hoffmansc) [reply aif360/sklearn/metrics/metrics.py]: Normally, if both `y_true` and `y_pred` are provided, it computes the `difference(selection_rate, ...))` and `selection_rate` specifies that `y_true` is ignored (just `y_pred` is used).

The `map()` function will index all of the `y_*` args, in this case `y_true` *and* `y_pred` (but, again, only `y_pred` is used by `selection_rate`).

File: aif360/sklearn/postprocessing/calibrated_equalized_odds.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+        for grp_idx in range(2):
+            i = (groups == self.groups_[grp_idx])
+            to_replace = (rng.rand(sum(i)) < self.mix_rates_[grp_idx])
+            new_preds = y_pred[i].copy()
+            new_preds[to_replace] = self.base_rates_[grp_idx]
+            yt[i] = new_preds
+
+        return np.c_[1 - yt, yt] if pos_idx == 1 else np.c_[yt, 1 - yt]
+
+    def predict(self, y_pred):
```
[the remark this answers is not in the document]
Comment (hoffmansc) [reply aif360/sklearn/postprocessing/calibrated_equalized_odds.py]: But the classifier is supposed to be calibrated so 0.5 should be the best threshold.

File: aif360/sklearn/postprocessing/__init__.py [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+from sklearn.base import BaseEstimator, MetaEstimatorMixin, clone
+from sklearn.model_selection import train_test_split
+from sklearn.utils.metaestimators import if_delegate_has_method
+from sklearn.utils.validation import check_is_fitted
+
+from aif360.sklearn.postprocessing.calibrated_equalized_odds import CalibratedEqualizedOdds
+
+
+class PostProcessingMeta(BaseEstimator, MetaEstimatorMixin):
+    """A meta-estimator which wraps a given estimator with a post-processing
```
[the remark this answers is not in the document]
Comment (hoffmansc) [reply aif360/sklearn/postprocessing/__init__.py]: This is a good point. I can add a 'prefit' option which doesn't train the estimator and uses all the data to train the post-processor.

If the estimator does not follow sklearn, it is possible to use the `CalibratedEqualizedOdds` class directly it just won't work in a Pipeline. You would just `fit()` and `predict()` using the probability scores from the base estimator.

Currently, the sklearn Pipeline expects only one estimator (`predict*` step) at the end of the pipeline so this is the only way I think it works but the meta-estimator itself can be part of a Pipeline, no problem.

File: examples/sklearn/demo_new_features.ipynb [hunk 9]
```diff
@@ -0,0 +1,782 @@
+{
```
[the remark this answers is not in the document]
Comment (hoffmansc) [reply examples/sklearn/demo_new_features.ipynb]: Regarding the last question: since the data is just a DataFrame now, the easiest way would probably be to add a new level to the index (e.g. `df.set_index([...], append=True)`). I could add an example to the notebook too.

**Comment by animeshsingh** [comment]
cc @adrinjalali 

Adrian this is a PR originating from the original request
https://github.com/IBM/AIF360/issues/58

Would be great to get your feedback and review on this, as well as how can we target this toward SKLearn community?

**Comment by animeshsingh** [comment]
@hoffmansc @nrkarthikeyan best to get this merged if the issues are non blockers, and then we can come back with refinements.
